extra_scripts/create_imagenet_a_data_files.py

```diff
@@ -0,0 +1,298 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
+
+import argparse
+import os
+from typing import Dict
+
+import numpy as np
+import torchvision.datasets as datasets
+from torchvision.datasets.utils import download_and_extract_archive
+from tqdm import tqdm
+
+
+def get_argument_parser():
+    """
+    List of arguments supported by the script
+    """
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "-i",
+        "--input",
+        type=str,
+        help="Path to the folder containing the original imagenet-a folder",
+    )
+    parser.add_argument(
+        "-o",
+        "--output",
+        type=str,
+        help="Folder where the classification dataset will be written",
+    )
+    parser.add_argument(
+        "-d",
+        "--download",
+        action="store_const",
+        const=True,
+        default=False,
+        help="To download the original dataset in the input folder",
+    )
+    return parser
+
+
+def download_dataset(root: str):
+    """
+    Download the Imagenet-R dataset archive and expand it
+    in the folder provided as parameter
+    """
+    URL = "https://people.eecs.berkeley.edu/~hendrycks/imagenet-a.tar"
+    download_and_extract_archive(url=URL, download_root=root)
+
+
+def create_imagenet_test_files(
+    input_path: str, output_path: str, target_mappings: Dict[int, int]
+):
+    """
+    Create a test split for ImageNet, by reading the image folder in the
+    'input_path' and mapping its targets to the targets of imagenet
+    using the 'target_mappings' dictionary
+    """
+
+    # Map the images of the image folder to their corresponding targets
+    # in ImageNet
+    image_paths = []
+    image_labels = []
+    imagenet_r = datasets.ImageFolder(root=input_path, loader=lambda x: x)
+    for image_path, target_id in tqdm(imagenet_r):
+        image_paths.append(image_path)
+        image_labels.append(target_mappings[target_id])
+
+    # Save the these lists in the disk_filelist format
```

Comment:
if the data is usable as a disk_folder, why do we need to convert it to disk_filelist? just curious :) I am sure you have a good reason :)

Reply:
Yes, I am not 100% confident about this either.

Here is the key problem I tried to solve doing this: the benchmark consists in training Imagenet and then evaluating on Imagenet-A or Imagenet-R.

But Imagenet-A and Imagenet-R do not have the same amount of classes as Imagenet, and have instead a sub-set of the classes (200 classes among the 1000 classes). In addition to this, the classes do not match the first 200, but are a random sample of the whole 1000 classes.

=> So, if I load the dataset with ImageFolder, I end up with 200 targets, and these targets will be in the range [0, 200) and will not match their corresponding Imagenet targets in the range [0, 1000).

So I thought of 2 options:
1. add empty folders to the Imagenet-A and Imagenet-R datasets, these folders would make sure that the same number of folders (with the same name) appear in both Imagenet and Imagenet-A/R
2. use the disk_filelist format to remap the indices of Imagenet-A and Imagenet-R to their corresponding targets in Imagenet

I found the option 2 more elegant and went for it, but elegance is not a very tangible measure, I fully concede that :)

Reply:
ah I see your point. I think given you care about the subsets, it makes sense to me to try to create the `disk_filelist`. Let's do this. :)